service/dap/server.go

```diff
@@ -550,58 +550,66 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
 		s.config.Debugger.WorkingDir = wdParsed
 	}
 
-	noDebug := false
-	if noDebugArg, ok := request.Arguments["noDebug"]; ok {
-		if v, ok := noDebugArg.(bool); ok && v { // noDebug == true
-			noDebug = true
-		}
-	}
-
-	var err error
-	if noDebug {
-		if s.noDebugProcess, err = s.launchWithoutDebug(program, targetArgs, s.config.Debugger.WorkingDir); err != nil {
+	if noDebug, _ := request.Arguments["noDebug"].(bool); noDebug {
+		cmd, err := s.startNoDebugProcess(program, targetArgs, s.config.Debugger.WorkingDir)
+		if err != nil {
 			s.sendErrorResponse(request.Request, FailedToLaunch, "Failed to launch", err.Error())
 			return
 		}
-	} else {
-		if s.debugger, err = debugger.New(&s.config.Debugger, s.config.ProcessArgs); err != nil {
-			s.sendErrorResponse(request.Request, FailedToLaunch, "Failed to launch", err.Error())
-			return
+		// Skip 'initialized' event.
+		s.send(&dap.LaunchResponse{Response: *newResponse(request.Request)})
+
+		// Then, block until the program terminates or is stopped.
+		if err := cmd.Wait(); err != nil {
+			s.log.Errorf("program exited: %v", err)
 		}
+
+		stopped := false
+		s.mu.Lock()
+		stopped = s.noDebugProcess == nil // if it was stopped, this should be nil.
```

Comment:
If you set it to nil only here, you could check Cmd.ProcessState here, couldn't you? And that would simplify the logic in stopNoDebugProcess too a bit.

Reply:
The program can terminate and we end up here because 1) the program exited by itself - due to normal termination or due to other signals (segfault, ...), or 2) the program was killed by Kill from stopNoDebugProcess. I don't see how to distinguish the two cases cleanly with Cmd.ProcessState when the program was killed by external processes.  In case of 2, we don't want to issue TerminatedEvent. Feel free to revisit/update it when the teardown logic & async request serving is settled down.